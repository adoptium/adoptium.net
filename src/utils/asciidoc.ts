import Asciidoctor from "@asciidoctor/core";
import fs from "fs";
import path from "path";
import { decode } from "html-entities";

interface AsciidoctorDocument {
  getSourceLocation(): { getPath(): string };
}

interface AsciidoctorReader {
  pushInclude(
    content: string,
    file: string,
    target: string,
    line: number,
    attributes: Record<string, unknown>
  ): void;
}

interface AsciidoctorAttributes {
  [key: string]: unknown;
}

// Initialize Asciidoctor processor
const asciidoctor = Asciidoctor();

/**
 * Custom Asciidoctor include processor that resolves partials
 */
class CustomIncludeProcessor {
  private basePath: string;

  constructor(basePath: string) {
    this.basePath = basePath;
  }

  process(
    doc: AsciidoctorDocument,
    reader: AsciidoctorReader,
    target: string,
    attributes: AsciidoctorAttributes
  ) {
    // Get current file directory to resolve relative includes correctly
    const currentDir = path.dirname(doc.getSourceLocation().getPath());

    let includeFile = target;

    // If target is a relative path, resolve it from current file directory
    if (!path.isAbsolute(target) && !target.startsWith("/")) {
      includeFile = path.resolve(currentDir, target);
    }

    // If not found, try from the content base path
    if (!fs.existsSync(includeFile)) {
      includeFile = path.join(this.basePath, target);
    }

    if (fs.existsSync(includeFile)) {
      const content = fs.readFileSync(includeFile, "utf8");
      return reader.pushInclude(content, includeFile, target, 1, attributes);
    } else {
      console.warn(`Include file not found: ${includeFile}`);
      return reader.pushInclude(
        `[WARNING] Include file not found: ${target}`,
        target,
        target,
        1,
        attributes
      );
    }
  }
}

/**
 * Process AsciiDoc content with proper include resolution
 */
export function processAsciiDoc(
  filePath: string,
  content: string,
  options: Record<string, unknown> = {}
) {
  const basePath = path.dirname(filePath);
  const registry = asciidoctor.Extensions.create();

  // Create an instance of our custom processor
  const customProcessor = new CustomIncludeProcessor(basePath);

  // Try using a simple approach first - just skip extension registration
  // and see if the document still processes correctly
  const skipExtensions = true;

  if (!skipExtensions) {
    // Register include processor using a factory function
    const IncludeProcessor = function () {};
    IncludeProcessor.prototype.handles = function () {
      return true;
    };
    IncludeProcessor.prototype.process = function (
      doc: AsciidoctorDocument,
      reader: AsciidoctorReader,
      target: string,
      attrs: AsciidoctorAttributes
    ) {
      return customProcessor.process(doc, reader, target, attrs);
    };

    registry.includeProcessor(IncludeProcessor);
  }

  // Set default options with proper typing
  const defaultOptions: Record<string, unknown> = {
    safe: "server",
    base_dir: basePath,
    attributes: {
      "source-highlighter": "highlight.js",
      icons: "font",
      sectanchors: "",
      idprefix: "",
      idseparator: "-",
    },
  };

  // Add extension registry only if we're not skipping extensions
  if (!skipExtensions) {
    defaultOptions.extension_registry = registry;
  }

  // Merge options
  const mergedOptions = { ...defaultOptions, ...options };

  // Convert AsciiDoc to HTML - use the processor without extension registration
  // if we're having issues with extensions
  try {
    let html = asciidoctor.convert(content, mergedOptions) as string;
    // Post-process image src paths: replace src="filename.png" with src="/docs/filename.png" if not already absolute or external
    html = html.replace(
      /<img([^>]+)src=["'](?![a-z]+:|\/)([^"'>]+)["']/gi,
      '<img$1src="/docs/$2"'
    );
    return html;
  } catch (error) {
    console.error("Error converting AsciiDoc content:", error);

    // Try without extension registry as fallback
    // Create a new options object without the extension_registry property using destructuring
    // Use object destructuring with "rest" syntax to remove the property
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { extension_registry, ...fallbackOptions } = defaultOptions;
    let html = asciidoctor.convert(content, fallbackOptions) as string;
    html = html.replace(
      /<img([^>]+)src=["'](?![a-z]+:|\/)([^"'>]+)["']/gi,
      '<img$1src="/docs/$2"'
    );
    return html;
  }
}

/**
 * Decode HTML entities in a string
 */
export function decodeHtmlEntities(text: string): string {
  return decode(text);
}

/**
 * Extract metadata from AsciiDoc document
 */
export function extractMetadata(
  filePath: string,
  content: string
): {
  title: string;
  description: string;
  authors: string[];
  attributes: Record<string, unknown>;
} {
  const doc = asciidoctor.load(content, { safe: "server" });

  // Get the title and ensure it's a string
  const docTitle = doc.getDocumentTitle();
  let title =
    typeof docTitle === "string"
      ? docTitle
      : docTitle?.toString() || path.basename(filePath, path.extname(filePath));

  // Decode HTML entities in the title
  title = decodeHtmlEntities(title);

  // Get and decode description if it exists
  const description = doc.hasAttribute("description")
    ? decodeHtmlEntities(String(doc.getAttribute("description")))
    : "";

  return {
    title,
    description,
    authors: doc.hasAttribute("page-authors")
      ? String(doc.getAttribute("page-authors"))
          .split(",")
          .map((author: string) => author.trim())
      : [],
    attributes: doc.getAttributes() as Record<string, unknown>,
  };
}

/**
 * Check if a file exists
 */
export function fileExists(filePath: string): boolean {
  try {
    return fs.existsSync(filePath);
  } catch {
    // Silently handle any errors and return false
    return false;
  }
}

/**
 * Get all available language variants for a given base path
 */
export function getLanguageVariants(
  basePath: string,
  fileName: string
): string[] {
  const dir = path.dirname(basePath);
  const baseFileName = path.basename(fileName, ".adoc");

  // Get all adoc files in the directory
  const files = fs
    .readdirSync(dir)
    .filter((file) => file.startsWith(baseFileName) && file.endsWith(".adoc"));

  // Extract locales from filenames
  return files.map((file) => {
    // Format: index.de.adoc -> de
    const match = file.match(
      new RegExp(`${baseFileName}(?:\\.([\\w-]+))?\\.adoc$`)
    );
    return match && match[1] ? match[1] : "en";
  });
}
