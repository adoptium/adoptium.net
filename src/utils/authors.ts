import AuthorData from "@/data/authors.json";

export interface Author {
  name: string;
  summary?: string;
  twitter?: string;
  github?: string;
  linkedin?: string;
}

export function getFormattedAuthorData(authorId: string): Author {
  const rawAuthorData = AuthorData[authorId as keyof typeof AuthorData];

  if (!rawAuthorData) {
    return {
      name: "Unknown Author",
    };
  }

  // Convert the raw author data to our Author interface
  return {
    name: rawAuthorData.name || "Unknown Author",
    // Convert null values to undefined
    summary: rawAuthorData.summary || undefined,
    twitter: rawAuthorData.twitter || undefined,
    github: rawAuthorData.github || undefined,
    linkedin: rawAuthorData.linkedin || undefined,
  };
}
