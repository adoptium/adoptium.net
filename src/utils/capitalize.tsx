// Format text with a capital letter
export function capitalize(text: string) {
  switch (text) {
    case "alpine-linux":
      return "Alpine Linux"
    case "alpine_linux":
      return "Alpine Linux"
    case "redhat":
      return "Red Hat"
    case "ibm":
      return "IBM"
    case "mac":
      return "macOS"
    case "semeru":
      return "Semeru Runtimes"
    default:
      return text.charAt(0).toUpperCase() + text.slice(1)
  }
}
