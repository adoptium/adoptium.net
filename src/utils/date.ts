export function formatDate(
  date: string = "",
  locale: string = "en-US"
): string {
  const targetDate = date ? new Date(date) : new Date();
  return targetDate.toLocaleDateString(locale, {
    month: "long",
    day: "2-digit",
    year: "numeric",
  });
}
