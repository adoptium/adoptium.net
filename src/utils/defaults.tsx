// This file is used to store defaults which can be accessed by other Javascript/Typescript functions

export const packageTypes = [
  {
    name: "JDK",
    value: "jdk",
  },
  {
    name: "JRE",
    value: "jre",
  },
]

const defaultPackageType = "jdk"
const defaultArchitecture = "x64"

export { defaultPackageType, defaultArchitecture }
