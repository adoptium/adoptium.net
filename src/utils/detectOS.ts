export enum UserOS {
  MAC = "MAC",
  WIN = "WIN",
  UNIX = "UNIX",
  LINUX = "LINUX",
  MOBILE = "MOBILE",
  UNKNOWN = "UNKNOWN",
}

export function detectOS(): UserOS {
  let OS = UserOS.UNKNOWN
  if (typeof window !== "undefined") {
    const appVersion = navigator.appVersion;
    if (appVersion.includes("Win")) {
      OS = UserOS.WIN
    } else if (appVersion.includes("Mac")) {
      OS = UserOS.MAC
    } else if (appVersion.includes("X11")) {
      OS = UserOS.UNIX
    } else if (appVersion.includes("Linux")) {
      OS = UserOS.LINUX
    } else if (appVersion.includes("Mobi")) {
      OS = UserOS.MOBILE
    }
  }
  return OS
}
