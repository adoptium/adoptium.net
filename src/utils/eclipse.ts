import axios from "axios"

const baseUrl = "https://newsroom.eclipse.org/api"

export async function fetchLatestNews() {
  const url = new URL(`${baseUrl}/news`)
  url.searchParams.append("parameters[publish_to]", "adoptium")
  url.searchParams.append("page", "1")
  url.searchParams.append("pagesize", "100")

  return axios
    .get(url.toString())
    .then(function (response) {
      return response.data
    })
    .catch(function () {
      return { news: [], pager: null }
    })
}
