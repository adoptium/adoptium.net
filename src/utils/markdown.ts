import fs from 'fs'
import path from 'path'
import matter from 'gray-matter'
import { unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'
import remarkRehype from 'remark-rehype'
import rehypeStringify from 'rehype-stringify'

/**
 * MDX options configuration for next-mdx-remote
 * This adds GitHub Flavored Markdown support (tables, task lists, strikethrough, etc)
 */
export const mdxOptions = {
  mdxOptions: {
    remarkPlugins: [remarkGfm],
    // You can add more plugins here if needed
  }
};

type Metadata = {
  title: string
  date: string
  author: string
  description: string
  tags?: string[]
  featuredImage?: string
}

function parseFrontmatter(fileContent: string) {
  // Use gray-matter to parse the frontmatter
  const { data, content } = matter(fileContent)
  
  // Ensure tags is always an array if present
  if (data.tags && !Array.isArray(data.tags)) {
    data.tags = [data.tags]
  }

  return { metadata: data as Metadata, content }
}

type MDXFile = {
  slug: string;
  path: string;
}

function getMDXFiles(dir: string): MDXFile[] {
  const results: MDXFile[] = []
  const files = fs.readdirSync(dir, { withFileTypes: true })
  
  for (const file of files) {
    if (file.isDirectory()) {
      // Check if index.md exists in this directory
      const indexPath = path.join(dir, file.name, 'index.md')
      if (fs.existsSync(indexPath)) {
        results.push({
          slug: file.name,
          path: indexPath
        })
      }
    } else if (file.isFile() && path.extname(file.name) === '.md') {
      // Support direct .md files too for flexibility
      results.push({
        slug: path.basename(file.name, '.md'),
        path: path.join(dir, file.name)
      })
    }
  }
  
  return results
}

function readMDXFile(filePath: string) {
  const rawContent = fs.readFileSync(filePath, 'utf-8')
  return parseFrontmatter(rawContent)
}

// Process markdown content with GitHub Flavored Markdown support
export async function processMarkdown(content: string): Promise<string> {
  const result = await unified()
    .use(remarkParse) // Parse markdown content
    .use(remarkGfm)   // Add GitHub Flavored Markdown support (tables, strikethrough, task lists, etc.)
    .use(remarkRehype) // Convert to HTML
    .use(rehypeStringify) // Stringify HTML
    .process(content)

  return result.toString()
}

function getMDXData(dir: string) {
  const mdxFiles = getMDXFiles(dir)
  return mdxFiles.map((file) => {
    const { metadata, content } = readMDXFile(file.path)

    // Parse the date components from date for URL construction
    let urlParts = { year: '', month: '' };
    // Default to current date if date is missing
    const date = metadata.date || new Date().toISOString().slice(0, 10);
    const [year = '', month = ''] = date.split('-');
    urlParts = { year, month };
    
    return {
      metadata,
      slug: file.slug,
      content,
      ...urlParts
    }
  })
}

export function getBlogPosts() {
  return getMDXData(path.join(process.cwd(), 'content', 'blog'))
}
