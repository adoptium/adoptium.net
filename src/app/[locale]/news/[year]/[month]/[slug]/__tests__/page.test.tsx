import { describe, it, expect, vi, afterEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import { axe } from 'vitest-axe';
import NewsArticlePage, { generateMetadata } from '../page';
import * as markdownUtils from '@/utils/markdown';
import * as dateUtils from '@/utils/date';
import { notFound } from 'next/navigation';

vi.mock('@/components/CustomMDX', () => ({
  CustomMDX: () => <div>MDX Content</div>
}));
vi.mock('@/utils/markdown');
vi.mock('@/utils/date');
vi.mock('next/navigation', () => ({ notFound: vi.fn() }));

const mockGetBlogPosts = markdownUtils.getBlogPosts as unknown as ReturnType<typeof vi.fn>;
const mockFormatDate = dateUtils.formatDate as unknown as ReturnType<typeof vi.fn>;

describe('NewsArticlePage', () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  it('renders article when post exists', async () => {
    mockGetBlogPosts.mockReturnValue([
      {
        slug: 'test-article',
        metadata: {
          title: 'Test Article',
          description: 'desc',
          date: '2025-06-18',
          author: 'Jane Doe',
          tags: ['temurin']
        },
        year: '2025',
        month: '06',
        content: 'Hello world'
      }
    ]);
    mockFormatDate.mockReturnValue('June 18, 2025');
    const element = await NewsArticlePage({ params: Promise.resolve({ slug: 'test-article', year: '2025', month: '06', locale: 'en' }) });
    const { container } = render(element);
    expect(screen.getByText('Test Article')).toBeInTheDocument();
    expect(screen.getByText('June 18, 2025')).toBeInTheDocument();
    // Snapshot test
    expect(container).toMatchSnapshot();
    // Axe accessibility lint
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('calls notFound if post does not exist', async () => {
    mockGetBlogPosts.mockReturnValue([]); // No posts at all
    await NewsArticlePage({ params: Promise.resolve({ slug: 'missing', year: '2025', month: '06', locale: 'en' }) });
    expect(notFound).toHaveBeenCalled();
  });

  it('generateMetadata returns correct metadata', async () => {
    mockGetBlogPosts.mockReturnValue([
      {
        slug: 'test-article',
        metadata: {
          title: 'Test Article',
          description: 'desc',
          date: '2025-06-18',
          author: 'Jane Doe',
          tags: ['temurin']
        },
        year: '2025',
        month: '06',
        content: 'Hello world'
      }
    ]);
    const meta = await generateMetadata({ params: Promise.resolve({ slug: 'test-article', year: '2025', month: '06', locale: 'en' }) });
    expect(meta).toBeDefined();
    expect(meta!.title).toBe('Test Article');
    expect(meta!.description).toBe('desc');
  });
});
