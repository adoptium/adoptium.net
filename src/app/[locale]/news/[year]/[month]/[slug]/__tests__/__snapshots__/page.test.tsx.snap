// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewsArticlePage > renders article when post exists 1`] = `
<div>
  <section>
    <script
      type="application/ld+json"
    >
      {"@context":"https://schema.org","@graph":[{"@type":"Person","name":"Jane Doe","description":"No biography available","affiliation":{"@type":"Organization","name":"Eclipse Adoptium"},"url":"https://adoptium.net/news/author/jdoe"},{"@type":"BlogPosting","headline":"Test Article","datePublished":"2025-06-18T00:00:00.000Z","description":"desc","image":"https://adoptium.net/news/2025/06/test-article/opengraph-image","url":"https://adoptium.net/news/2025/06/test-article","author":{"@type":"Person","name":"Jane Doe","url":"https://adoptium.net/news/author/jdoe"},"publisher":{"@type":"Organization","name":"Eclipse Adoptium","logo":{"@type":"ImageObject","url":"https://adoptium.net/images/adoptium-icon.png"}}}]}
    </script>
    <div
      class="pt-16 md:pt-24 pb-12"
    >
      <div
        class="mx-auto max-w-[832px] w-full px-6 lg:px-0 flex flex-col items-center justify-center"
      >
        <div
          class="self-stretch flex-col justify-center items-center gap-6 flex pb-4"
        >
          <div
            class="self-stretch flex-col justify-center items-center gap-4 flex"
          >
            <div
              class="justify-start items-center gap-3 inline-flex"
            >
              <svg
                fill="none"
                height="14"
                viewBox="0 0 14 14"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <rect
                  fill="#FF1464"
                  height="8"
                  rx="2"
                  width="8"
                  x="3"
                  y="3"
                />
                <rect
                  height="11"
                  rx="3.5"
                  stroke="#FF1464"
                  stroke-opacity="0.25"
                  stroke-width="3"
                  width="11"
                  x="1.5"
                  y="1.5"
                />
              </svg>
              <div
                class="text-rose-600 text-base font-bold leading-normal"
              >
                News article
              </div>
            </div>
            <h1
              class="self-stretch text-center text-white text-2xl md:text-5xl pb-4 font-semibold break-words"
            >
              <span
                class="break-words whitespace-normal"
              >
                Test Article
              </span>
            </h1>
          </div>
          <div
            class="self-stretch text-center text-grey text-grey-300 text-lg md:text-xl font-normal leading-7"
          >
            desc
          </div>
          <div
            class="flex justify-center items-center gap-5"
          >
            <div
              class="flex justify-center items-center gap-5"
            >
              <img
                alt="Jane Doe"
                class="ml-2 mb-0 rounded-full"
                data-nimg="1"
                decoding="async"
                height="30"
                loading="lazy"
                src="/_next/image?url=%2Fimages%2Fauthors%2Fjdoe.jpg&w=64&q=75"
                srcset="/_next/image?url=%2Fimages%2Fauthors%2Fjdoe.jpg&w=32&q=75 1x, /_next/image?url=%2Fimages%2Fauthors%2Fjdoe.jpg&w=64&q=75 2x"
                style="color: transparent;"
                width="30"
              />
              <span
                class="text-[16px] font-bold leading-[150%] text-white"
              >
                <a
                  href="/news/author/jdoe"
                >
                  <span
                    class="text-primary pr-2"
                  >
                    Jane Doe
                    ,
                  </span>
                   
                  June 18, 2025
                </a>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <section
      class="w-full mx-auto p-0 md:p-6 lg:px-0 flex flex-col md:items-center justify-center"
    >
      <div
        class="w-full max-w-full m-2 md:m-4 md:max-w-4xl bg-white/5 rounded-2xl shadow-2xl p-2 md:p-12 relative z-10 overflow-x-auto"
      >
        <article>
          <header
            class="py-5 flex flex-col md:flex-row md:items-center md:justify-between gap-4 border-b border-white/10 mb-8 px-2 md:px-0"
          >
            <div
              id="share-buttons"
            >
              <button
                aria-label="Share on X (Twitter)"
                class="react-share__ShareButton mr-2"
                style="background-color: transparent; padding: 0px; cursor: pointer;"
              >
                <svg
                  fill="currentColor"
                  height="25"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M389.2 48h70.6L305.6 224.2 487 464H345L233.7 318.6 106.5 464H35.8L200.7 275.5 26.8 48H172.4L272.9 180.9 389.2 48zM364.4 421.8h39.1L151.1 88h-42L364.4 421.8z"
                  />
                </svg>
              </button>
              <button
                aria-label="Share on Linkedin"
                class="react-share__ShareButton mr-2"
                style="background-color: transparent; padding: 0px; cursor: pointer;"
              >
                <svg
                  fill="currentColor"
                  height="25"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"
                  />
                </svg>
              </button>
              <button
                aria-label="Share on Facebook"
                class="react-share__ShareButton mr-2"
                style="background-color: transparent; padding: 0px; cursor: pointer;"
              >
                <svg
                  fill="currentColor"
                  height="25"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M504 256C504 119 393 8 256 8S8 119 8 256c0 123.78 90.69 226.38 209.25 245V327.69h-63V256h63v-54.64c0-62.15 37-96.48 93.67-96.48 27.14 0 55.52 4.84 55.52 4.84v61h-31.28c-30.8 0-40.41 19.12-40.41 38.73V256h68.78l-11 71.69h-57.78V501C413.31 482.38 504 379.78 504 256z"
                  />
                </svg>
              </button>
              <button
                aria-label="Share on Reddit"
                class="react-share__ShareButton mr-2"
                style="background-color: transparent; padding: 0px; cursor: pointer;"
              >
                <svg
                  fill="currentColor"
                  height="25"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M201.5 305.5c-13.8 0-24.9-11.1-24.9-24.6 0-13.8 11.1-24.9 24.9-24.9 13.6 0 24.6 11.1 24.6 24.9 0 13.6-11.1 24.6-24.6 24.6zM504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-132.3-41.2c-9.4 0-17.7 3.9-23.8 10-22.4-15.5-52.6-25.5-86.1-26.6l17.4-78.3 55.4 12.5c0 13.6 11.1 24.6 24.6 24.6 13.8 0 24.9-11.3 24.9-24.9s-11.1-24.9-24.9-24.9c-9.7 0-18 5.8-22.1 13.8l-61.2-13.6c-3-.8-6.1 1.4-6.9 4.4l-19.1 86.4c-33.2 1.4-63.1 11.3-85.5 26.8-6.1-6.4-14.7-10.2-24.1-10.2-34.9 0-46.3 46.9-14.4 62.8-1.1 5-1.7 10.2-1.7 15.5 0 52.6 59.2 95.2 132 95.2 73.1 0 132.3-42.6 132.3-95.2 0-5.3-.6-10.8-1.9-15.8 31.3-16 19.8-62.5-14.9-62.5zM302.8 331c-18.2 18.2-76.1 17.9-93.6 0-2.2-2.2-6.1-2.2-8.3 0-2.5 2.5-2.5 6.4 0 8.6 22.8 22.8 87.3 22.8 110.2 0 2.5-2.2 2.5-6.1 0-8.6-2.2-2.2-6.1-2.2-8.3 0zm7.7-75c-13.6 0-24.6 11.1-24.6 24.9 0 13.6 11.1 24.6 24.6 24.6 13.8 0 24.9-11.1 24.9-24.6 0-13.8-11-24.9-24.9-24.9z"
                  />
                </svg>
              </button>
              <button
                aria-label="Share on Bluesky"
                class="react-share__ShareButton mr-2"
                style="background-color: transparent; padding: 0px; cursor: pointer;"
              >
                <svg
                  fill="currentColor"
                  height="25"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 576 512"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M407.8 294.7c-3.3-.4-6.7-.8-10-1.3c3.4 .4 6.7 .9 10 1.3zM288 227.1C261.9 176.4 190.9 81.9 124.9 35.3C61.6-9.4 37.5-1.7 21.6 5.5C3.3 13.8 0 41.9 0 58.4S9.1 194 15 213.9c19.5 65.7 89.1 87.9 153.2 80.7c3.3-.5 6.6-.9 10-1.4c-3.3 .5-6.6 1-10 1.4C74.3 308.6-9.1 342.8 100.3 464.5C220.6 589.1 265.1 437.8 288 361.1c22.9 76.7 49.2 222.5 185.6 103.4c102.4-103.4 28.1-156-65.8-169.9c-3.3-.4-6.7-.8-10-1.3c3.4 .4 6.7 .9 10 1.3c64.1 7.1 133.6-15.1 153.2-80.7C566.9 194 576 75 576 58.4s-3.3-44.7-21.6-52.9c-15.8-7.1-40-14.9-103.2 29.8C385.1 81.9 314.1 176.4 288 227.1z"
                  />
                </svg>
              </button>
              <button
                aria-label="Share via Email"
                class="react-share__ShareButton mr-2"
                style="background-color: transparent; padding: 0px; cursor: pointer;"
              >
                <svg
                  fill="currentColor"
                  height="25"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="25"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M48 64C21.5 64 0 85.5 0 112c0 15.1 7.1 29.3 19.2 38.4L236.8 313.6c11.4 8.5 27 8.5 38.4 0L492.8 150.4c12.1-9.1 19.2-23.3 19.2-38.4c0-26.5-21.5-48-48-48L48 64zM0 176L0 384c0 35.3 28.7 64 64 64l384 0c35.3 0 64-28.7 64-64l0-208L294.4 339.2c-22.8 17.1-54 17.1-76.8 0L0 176z"
                  />
                </svg>
              </button>
            </div>
            <div
              class="flex flex-wrap gap-2 mt-6"
            >
              <a
                class="px-3 py-1 text-sm bg-pink hover:bg-rose-800 transition-colors rounded-full text-white"
                href="/news/tag/temurin"
              >
                #
                temurin
              </a>
            </div>
          </header>
          <article
            class="prose prose-invert lg:prose-lg max-w-full px-2 md:px-0"
          >
            <div>
              MDX Content
            </div>
          </article>
        </article>
        <div
          class="mt-8 px-2 md:px-0"
        >
          <ul
            class="flex flex-wrap justify-between list-none pt-5 md:p-0"
          >
            <li />
            <li />
          </ul>
        </div>
      </div>
    </section>
  </section>
</div>
`;
